src/main/java/roomescape/service/ReservationService.java

```diff
@@ -0,0 +1,62 @@
+package roomescape.service;
+
+import java.util.List;
+import org.springframework.stereotype.Service;
+import roomescape.domain.Reservation;
+import roomescape.domain.ReservationTime;
+import roomescape.domain.Theme;
+import roomescape.dto.ReservationResponse;
+import roomescape.dto.ReservationSaveRequest;
+import roomescape.exception.AlreadyExistReservationException;
+import roomescape.exception.IllegalThemeException;
+import roomescape.exception.IllegalTimeException;
+import roomescape.mapper.ReservationMapper;
+import roomescape.repository.ReservationDao;
+import roomescape.repository.ThemeDao;
+import roomescape.repository.TimeDao;
+
+@Service
+public class ReservationService {
+
+    private final ReservationMapper reservationMapper = new ReservationMapper();
+    private final ReservationDao reservationDao;
+    private final TimeDao timeDao;
+    private final ThemeDao themeDao;
+
+    public ReservationService(final TimeDao timeDao, ReservationDao reservationDao, ThemeDao themeDao) {
+        this.timeDao = timeDao;
+        this.reservationDao = reservationDao;
+        this.themeDao = themeDao;
+    }
+
+    public List<ReservationResponse> findAllReservations() {
+        List<Reservation> reservations = reservationDao.findAll();
+        return reservations.stream()
+                .map(reservationMapper::mapToResponse)
+                .toList();
+    }
+
+    public ReservationResponse saveReservation(ReservationSaveRequest request) {
+        if (request.timeId() == null) {
+            throw new IllegalTimeException("[ERROR] 유효하지 않은 형식의 예약 시간입니다.");
+        }
+        ReservationTime time = timeDao.findById(request.timeId());
+
+        if (request.themeId() == null) {
+            throw new IllegalThemeException("[ERROR] 유효하지 않은 형식의 테마입니다.");
+        }
+        Theme theme = themeDao.findById(request.themeId());
+        Reservation reservation = reservationMapper.mapToReservation(request, time, theme);
+
+        if (reservationDao.existByDateTimeTheme(reservation.getDate(), time.getStartAt(), reservation.getThemeId())) {
+            throw new AlreadyExistReservationException("[ERROR] 같은 날짜, 테마, 시간에 중복된 예약을 생성할 수 없습니다.");
+        }
+        Long saveId = reservationDao.save(reservation);
```

Comment:
여기는 중요한 건 아니지만 작성된 로직을 기준으로 본다면 아래와 같은 흐름이 더 자연스러운 것 같아요. 저장을 하기 위한 Reservation 객체를 만든 후 존재 여부를 확인하기보단 저장할 데이터에 대한 존재 여부를 확인 후 저장하기 위한 객체를 만드는 것이 어떨까요?

```suggestion
        Theme theme = themeDao.findById(request.themeId());
        if (reservationDao.existByDateTimeTheme(request.getDate(), time.getStartAt(), theme.getId())) {
            throw new AlreadyExistReservationException("[ERROR] 같은 날짜, 테마, 시간에 중복된 예약을 생성할 수 없습니다.");
        }
        
        Reservation reservation = reservationMapper.mapToReservation(request, time, theme);
        Long saveId = reservationDao.save(reservation);
```

추가로 시간도 또한 id로 확인이 가능할 것 같은데 id가 아닌 시간 값으로 확인하신 이유가 있을까요?

Reply:
```java
    public ReservationResponse saveReservation(ReservationRequest request) {
        ReservationTime time = timeDao.findById(request.timeId());

        if (request.date().isEqual(LocalDate.now()) && time.inPast()) {
            throw new IllegalReservationException("[ERROR] 과거 시간은 예약할 수 없습니다.");
        }

        Theme theme = themeDao.findById(request.themeId());
        Reservation reservation = reservationMapper.mapToReservation(request, time, theme);

        if (reservationDao.existByDateAndTimeAndTheme(reservation.getDate(), time.getStartAt(), reservation.getThemeId())) {
            throw new ExistReservationException("[ERROR] 같은 날짜, 테마, 시간에 중복된 예약을 생성할 수 없습니다.");
        }

        Reservation newReservation = reservationDao.save(reservation);
        return reservationMapper.mapToResponse(newReservation);
    }
```
리팩토링한 현재 버전의 경우 Reservation 객체 생성이 우선되어야한다고 생각합니다!
왜냐하면 Reservation에서 예약자 이름, 날짜가 blank인 경우를 검증합니다. 따라서 서비스에 있는 중복 예약인지 검증은 이러한 빈값 검증 이후에 일어나는 것이 적절하다고 생각하여 해당 순서로 배치해보고자 합니다!

> 추가로 시간도 또한 id로 확인이 가능할 것 같은데 id가 아닌 시간 값으로 확인하신 이유가 있을까요?